Add support for importing authority data from KulturNav

hemed opened the pull request:
This PR adds support for importing authority data from [KulturNav](https://kulturnav.org/) - a popular authority register in Nordic, especially in Norway and Sweden. 

This feature allows users to search for the authority data from KulturNav and import them into Goobi based on the predefined settings. It comes as one additional authority source beside `viaf,` `easydb` and `GND` in the metadata editor. The feature is configurable, for example, one can restrict searching only to a specific dataset. See the examples on <code>goobi_metadataDisplayRules.xml</code>. If nothing is specified in the `<source>` element, it will search for the entire knowledgebase in KulturNav. 

The integration is very important for us at the University of Bergen Library and I believe it will be useful for other institutions using Goobi in the Scandinavia.  @SteffenHankiewicz please test and let me know if any.  We would be happy to see this integrated into Goobi workflow.

hemed committed:
Add development version of Goobi workflow

hemed committed:
Add plugins to git for easy debugging

Commit:
Start to implement data importer

hemed committed:
Continue implementing KulturNav importer and make changes to docker-compose file

hemed committed:
Implement generic JSON importer, create some utility methods to manipulate norm data and parse search terms before hitting KulturNav endpoint

hemed committed:
Customize frontend and adapt the backend to support the integration with KulturNav

hemed committed:
Work with integration to MetaPerson and related forms

Commit:
Add import icons and make some changes to JSF forms

hemed committed:
Clear search query after executing search and enable searching using Enter

Commit:
Add new fields to the ruleset for testing purpose and do some code refactoring

hemed committed:
Read source for persons from config file goobi_metadataDisplayRules.xml so that users can specify

hemed committed:
Use svg icons, remeve extra top line from modal and restrict search to only Samla datasets

hemed committed:
Use svg icons, remove extra <hr> tag from modal and restrict search to only Samla datasets

hemed committed:
Add line break to the modal text

hemed committed:
Merge upstream master local kulturnav feature branch

hemed committed:
Resolve merge conflicts

hemed committed:
Resolve merge conflicts

Commit:
Resolve errors caused by merging

hemed committed:
Use button submit in the kulturnav modal

hemed committed:
Add translate labels to messages_de.properties

hemed committed:
Merge branch 'master' of github.com:intranda/goobi-workflow

hemed committed:
Add plugins folder to gitignore

hemed committed:
Remove plugins folder and the development version of docker-compose

hemed committed:
Remove the development version of docker-compose

hemed committed:
Merge the fature for kulturnav integration to develop branch

hemed committed:
Remove specific development configs

hemed committed:
Add example for usage of kulturnav integration

SteffenHankiewicz commented:
This does look very good. We will take a deeper look into this the following week and then come back to you and do the merge. Thank you so much for providing this implementation work to the Goobi community.

opaetzel commented on Goobi/Dockerfile:
```diff
@@ -41,6 +41,8 @@ RUN apt-get update && \
         rename \
         file \
+        vim \
+        unzip \
```
I don't think changing the Dockerfile is needed for the new feature. I would prefer it, if the docker image would stay as small as possible

opaetzel commented on Goobi/install/config/goobi_config.properties:
```diff
@@ -56,3 +56,4 @@ SsoHeaderName=YourHeaderName
 # This also affects "Calculate number of metadata and images" in process_all.xhtml
 # comma-separated list of suffixes. Default until now was just .tif
 historyImageSuffix=.jp2,.tif
+jwtSecret = 7zfnYtL16rY0U7N7LJBA
```
The jwtSecret will be randomly created by Goobi in the current version. But only if it doesn't exist yet. Having it configured in the default docker image doesn't look like a good idea.

opaetzel commented on Goobi/install/config/goobi_metadataDisplayRules.xml:
```diff
@@ -17,6 +17,18 @@
                 <label></label>
             </textarea>
+
+            <!-- Restrict to specific dataset, try to search for e.g Per -->
+            <kulturnav ref="person">
+                <source>entity.dataset:e0c682f9-a594-4fe1-8c01-f208f52d79ac</source>
+                <field></field>
+            </kulturnav>
+
+            <!-- Get place of publication from KulturNav -->
+            <kulturnav ref="PlaceOfPublication">
+                <source>entityType:Place</source>
+                <field></field>
+            </kulturnav>
```
I would also prefer a basic configuration in the docker image

hemed replied:
I didn't get here what you mean by basic configuration in the docker image. My understanding is that docker is copying everything from `Goobi/install/config` when building new image in which, this particular config will always be copied?

opaetzel replied:
Yes, it is copied to the docker image. So the person and PlaceOfPublication metadata will be a "kulturnav field" for the basic docker image. I would prefer to have these lines removed from the configuration, so we would have a basic configuration that just shows input fields for the metadata.

hemed replied:
This commit https://github.com/intranda/goobi-workflow/pull/23/commits/475d40cd80d8c2259e89c634911064c4b01fadfd removes the lines to have a basic configuration.

opaetzel commented on Goobi/install/config/messages_de.properties:
```diff
@@ -1 +0,0 @@
-info = Goobi Installation der CHANGEME in CHANGEME. <br /> <br /> F\u00FCr Supportanfragen wenden Sie sich bitte an: <a href="mailto:support@intranda.com">support@intranda.com</a>.
```
Why was this deleted? It is OK to add new messages, but please don't remove existing ones

opaetzel commented on Goobi/install/config/messages_en.properties:
```diff
@@ -1 +0,0 @@
-info = Goobi instance of CHANGEME CHANGEME. <br /> <br /> For support questions please ask: <a href="mailto:support@intranda.com">support@intranda.com</a>.
```
Why was this deleted? It is OK to add new messages, but please don't remove existing ones

opaetzel reviewed (changes requested):
I found a few changes to the default docker images that I would like to have reverted, everything else looks OK to me

hemed committed:
Revert messages_de.properties

hemed committed:
Revert changes from Dockerfile, messages and goobi_config

hemed committed:
Resolve conflicts for messages_de.properties

hemed commented:
Hi @opaetzel . Thanks for the review. I don't know why I deleted those messages from translation files, sorry for that. 
I have now reverted the changes. See https://github.com/intranda/goobi-workflow/pull/23/commits/ba8c661511932ea0450991e9611190ced55a234b

hemed committed:
Get Viaf to work and resolve an issue where Enter triggers row selection in Kulturnav modal

opaetzel commented on Goobi/webapp/uii/includes/metseditor/modals/inc_me_modal_kulturnav.xhtml:
```diff
@@ -0,0 +1,169 @@
+                            </div>
+                            <div class="col-2 text-right">
+                                <div class="text-right">
+                                    <h:commandLink
```
Please use buttons with jsf:action. This needs to be done to keep WCAG2 accessiblity. Anchors (`<h:commandLink>`) have a different meaning than a button. So we need to be using buttons for everything that triggers an action.

hemed replied:
I noticed some strange behavior when using a button - when you use "Enter" to search (for the second time) it does not trigger search but rather it automatically selects the first row of the results and close the modal. I will investigate more on this.

opaetzel replied:
oh, yes. You will need to set the attribute `type="button"`, otherwise it will be a submit-button and be the default action for the form. Once you have a button like this: `<button type="button">`, everything should work as expected

hemed replied:
Thanks for the tip. This is now resolved. I found out I was using `<button type="button">` for search button but  `<button type="submit">` for select record button.  Removing those (hence falling back to `<button type="submit">`) solved for issue.  See commit:  https://github.com/intranda/goobi-workflow/pull/23/commits/4fe66f0d98027f95da67b0b37e2659306239fa1b

opaetzel commented on Goobi/webapp/uii/includes/metseditor/modals/inc_me_modal_kulturnav.xhtml:
```diff
@@ -0,0 +1,169 @@
+
+
+                                        <div class="col-1">
+                                            <h:commandLink
```
Also use a `<button>` here, please

hemed committed:
Use basic configuration for goobi_metadataDisplayRules.xml

hemed committed:
Use buttons instead of commandLink in Kulturnav modal

opaetzel commented:
Hi @hemed, I made a new release on Friday, so there are some files with conflicts. Could you merge the current develop into your develop? I will then merge the changes and they will be part of the next release

hemed committed:
Merge remote-tracking branch 'upstream/develop' into develop

hemed committed:
Merge branch 'develop' of github.com:intranda/goobi-workflow into develop

hemed commented:
Hi @opaetzel, I have now merged the upstream/develop into my develop as you suggested.
